### ISSUE TITLE
Event handlers aren't traced

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#7781]
As I found out in #7780, when `fish_trace` is in effect variable handlers (and presumably other event handlers) don't show up.

E.g. (`>_` is my prompt, `>` is the trace and the line without a symbol is ordinary output):

```fish
>_ function foo --on-variable foo; echo foo handler; end
>_ set -g fish_trace 1
>_ set -g foo bar
> set -g foo bar
foo handler
```

It might be cool to show those as well, ideally somehow marked as event handlers. Maybe something like

```fish
>_ function foo --on-variable foo; echo foo handler; end
>_ set -g fish_trace 1
>_ set -g foo bar
> set -g foo bar
(event $foo) > echo foo handler
foo handler
```

?

**Comment by mqudsi** [comment]
Note that they're intentionally not traced at the moment, in the original commit that introduced `fish_trace` (a7f1d2c) @ridiculousfish writes

>  The user's prompt is not traced unless you run it explicitly. Events are not traced because it is noisy; however autoloading is.

It seems the concern was too much potentially irrelevant output rather than confusing intermixing of standard and event-sourced trace messages.

**Comment by faho** [comment]
Well, $fish_trace is for debugging, and right now it's quite awkward to not have a way to see *everything* that actually runs.

I would either enable events, or give a setting that enables them - set $fish_trace to 2 or "with-events" or something.

**Comment by CodingAnarchy** [comment]
In the case of #7780, the problem was the result of an event handler brought in by `asdf-java`. Without being able to trace that in the event handler, debugging was not obvious.

**Comment by ridiculousfish** [comment]
Yeah, fish currently checks if `$fish_trace` is non-empty but otherwise ignores its value. The plan was to later allow the value to control what gets traced, and the format of the output (e.g. maybe show line numbers). Maybe the simplest way to start is just with levels - 1 and 2 as faho proposes.

**Comment by faho** [comment]
Numeric levels here seem kind of awkward. They remind me of the numbered debug levels.

I'd probably go with something like flog categories - `set fish_trace normal` would set it as it is currently, but `set fish_trace normal,events` would add events to that. I probably wouldn't go overboard and allow `set fish_trace normal,event-variable-PWD` to *just* trace the variable handler for $PWD.

Unless we have absolutely no idea what else would ever be added for this, in which case either "1" and "2" or "normal" and "with-events" are okay.

**Comment by faho** [comment]
Okay, places where we suppress traces:

- Events
- The title
- The prompt
- While executing key bindings

We could enable all of these separately, or as one, or maybe don't enable the key bindings one - that seems noisy and likely to lead to graphical glitches.
